src/plat/bcm2711/overlay-rpi4.dts

```diff
@@ -18,35 +18,60 @@
 		    &{/timer};
 	};
 
+	/*
+	 * See BCM2711 TRM section 1.2 and
+	 * https://github.com/raspberrypi/firmware/issues/1374 for
+	 * reference.
+	 *
+	 * RPi4B seems to be in "low peripheral" mode by default.
+	 * These values were observed from running
+	 * Raspbian instance from /proc/iomem and /sys/firmware/devicetree/base
+	 *
+	 * By default, the RPi 4B has 76MB reserved for the VideoCore/GPU.
+	 * So, we have following memory ranges available:
+	 *
+	 * VC GPU RAM 76MB -> 0x40000000-0x04C00000 = 0x3b400000
+	 *
+	 * 0x00 0x00000000 0x3b400000  948MB
+	 * 0x00 0x40000000 0xbc000000  3008MB
+	 * 0x01 0x00000000 0x80000000  2048MB
+	 * 0x01 0x80000000 0x80000000  2048MB
+	 *
+	 * Disjunct memory areas seem to work only
+	 * when declared in different memory DT nodes.
+	 *
+	 * If all of the declarations below are in the
+	 * same node, the kernel build system seems
+	 * to only understand the second 3008MB area
+	 * and ignores the rest.
+	 */
+
+	/delete-node/ memory;
+
 	memory@0 {
-        /*
-         * See BCM2711 TRM section 1.2 and
-         * https://github.com/raspberrypi/firmware/issues/1374 for
-         * reference.
-         */
+		device_type = "memory";
+		reg = < 0x00000000 0x00000000 0x3b400000 >;
+	};
+
+	memory@40000000 {
+		device_type = "memory";
+		reg = < 0x00000000 0x40000000 0xbc000000 >;
+	};
 
-         /* RPi4B seems to be in "low peripheral" mode by default.
-          * These values were observed from running
-          * Raspbian instance from /proc/iomem and /sys/firmware/devicetree/base
-          *
-          * Set gpu_mem=64 so we have following memory ranges available
-          * VC GPU RAM 64MB -> 0x40000000-0x04000000 = 0x3c000000
-          *
-          * 0x00 0x00000000 0x3c000000  960MB
-          * 0x00 0x40000000 0xbc000000  3008MB
-          * 0x01 0x00000000 0x80000000  2048MB
-          * 0x01 0x80000000 0x80000000  2048MB
-          */
-        reg = < 0x00 0x00000000 0x3c000000 /* 960MB */
-                0x00 0x40000000 0xbc000000 /* 3008MB */
-                0x01 0x00000000 0x80000000 /* 2048MB */
-                0x01 0x80000000 0x80000000 /* 2048MB */ >;
```

Comment:
> Second fix is separating the disjunct memory areas
to separate nodes, as the generator Python scripts
in 'kernel/tools/' etc discarded the first area below
1GB limit if all areas were defined in the same node.
The effect of different variations can be observed
during kernel build in the generated
'kernel/gen_headers/plat/machine/devices_gen.h' header.

I don't think I understand what the error here is? When I generate the header file using this region definition the output I see generated appears to contain all of the regions (using b8ef00e09a1273d9f436dc67db4761cd4711295f):
```
static const p_region_t BOOT_RODATA avail_p_regs[] = {
    /* /memory@0 */
    {
        .start = 0x0,
        .end   = 0x3c000000
    },
    /* /memory@0 */
    {
        .start = 0x40000000,
        .end   = 0xfc000000
    },
    /* /memory@0 */
    {
        .start = 0x100000000,
        .end   = 0x200000000
    },
};
```

Reply:
Seems like you are right about that the `avail_p_regs` is generated correctly without having the memory areas in separate nodes. I had the issue occurring at some point in time, from where this observation and workaround was born from, but honestly it's been so long that I don't recall the exact circumstances for it. From what I recall, the last area

 `    /* /memory@0 */
    {
        .start = 0x100000000,
        .end   = 0x200000000
    }`

was being discarded by the build system for some reason.
Obviously, the behavior noted in the comment is no longer true, I will update the file accordingly.

Reply:
Fixed in 59be165.